### ISSUE TITLE
Post processing: the return link from Butterfly to Honeybee

### ISSUE THREAD

**Issue by TheodoreGalanos** [mostaphaRoudsari/Butterfly#12]
So I had a small discussion with Chris a few days back and I promised I would look into this. The whole thing circled around me wanting something and Chris being a cool guy and helping me do it. I wanted to add my CFD results to his comfort map workflow. We decided the first step would be to allow for the indication of an average velocity (from CFD) in the outdoor comfort calculations.

The extraction of these things can be done in various ways in Openfoam. One way is by post processing through parafoam. From what I have seen there are a number of filter pipelines that can extract selections into tabulated format, which I'm guessing we can then use through python to feed into our mesh in HB.

But I think I found a better way, at least a faster and more automated way that we can include in Butterfly itself. It is through the sampleDict utility (https://github.com/OpenFOAM/OpenFOAM-3.0.x/blob/master/applications/utilities/postProcessing/sampling/sample/sampleDict). 

This utility gives a number of options on where to sample from (face, cells, points, zones, patches, surfaces, etc.) and what to get as samples (Temperature, Velocity, Pressure, etc.). It seems to be the most suitable especially in the case of comfort simulations where usually large and complex surfaces are in play. sampleDict allows us to extract these complex regions just at the last timestep of the simulation.

I played around briefly however I can tell you the available formats it supports as an output:

-csv
-ensight
-gnuplot
-jplot
-raw
-vtk
-xmgr

The most suitable seems to be csv, although that is just to my eyes and not to coding requirements. The csv output looks something like this (extract of velocity from a face):

x	                     y	        z	                 U_0	               U_1   	       U_2
0	                 0.218	0	                 0	                 0	                  0 
0.00217143	0.218	4.32392E-08	0.000148327	-0.0726747	0.000311878
0.00434286	0.218	4.31451E-08	-3.95837E-05	-0.0987746	0.000508977
0.00651429	0.218	4.3051E-08	-0.000285569	-0.0930806	0.000553867
0.00868571	0.218	0.000000043	-0.000442578	-0.0747404	0.000501759
0.0108571	0.218	4.28629E-08	-0.000568711	-0.0543903	0.000387795
0.0130286	0.218	4.27688E-08	-0.000667924	-0.0341085	0.000248181
0.0152	0.218	4.26747E-08	-0.000763246	-0.014714	0.00011728
0.0173714	0.218	4.25806E-08	-0.000851649	0.00256538	0.000017424
0.0195429	0.218	4.24865E-08	-0.000920814	0.015989	-4.53957E-05

It's quite a straightforward output and an easy math calculation to get average wind velocity at these points. Notice this is a face, so it the points are actually quite close to eachother. I still haven't used it in a surface so I'm not sure of how we can actually control the point numbers (in order to align it with the mesh points component in HB). I will test more, especially with the other file formats and will forward the files when I have them for testing.

Kind regards,
Theodore.

**Comment by TheodoreGalanos** [comment]
Passing data was actually much simpler than anticipated.

The file format selected is csv, which is easily extracted by Paraview. Chris then developed/updated a couple of amazing components. This now allows us to do two important analyses:

- outdoor thermal comfort assessments, using Chris's known workflow and with the addition of actual CFD simulated wind speeds

- pedestrian/outdoor wind comfort studies. Thanks to a nifty small component that takes wind factors for each point (linked to a point in the viewfactormesh) extracted from Paraview and uses that information to create a prediction of yearly wind velocities, we can now very quickly (in matter of minutes really) conduct pedestrian wind comfort studies, calculating wind exceedance probabilities.

Quite interesting stuff I think and a nice showcase on how Butterfly would really link so well with the beast that is LB/HB.

More soon I hope!

Kind regards,
Theodore.

-

**Comment by mostaphaRoudsari** [comment]
Very exciting stuff! I'm keeping an eye on all your posts and still waiting for couple of days to spend on butterfly. The other workflow that you have found looks promising #13

**Comment by Vyrwn** [comment]
Hello,

this is very interesting. We have been working on the exact same workflow in our office. Getting the wind data into GH via a .csv file and use a factor for each grid. 

I am very interested in the method you used in ladybug. We run the results for a single day because we worked with a much larger grid and there were some complications with the indices component. Correct me if I am missing something but we could not run the analysis for the whole grid using the _Ladybug_Thermal_Comfort_Indices_ component. We had to "trick" the component to do so, using the number of the grid points instead of 8760 hours. 

It will be very interesting to see the method you used for the annual run, which seems very well optimized in order to run so fast.

keep up the nice work.

kind regards,
Byron

**Comment by chriswmackey** [comment]
I agree that this is all very exciting!

Byron,
I just added a microclimate map recipe for PET last night as I realized that both Theodore and a number of other people were interested in interacting with the PET model spatially:
![petmap](https://cloud.githubusercontent.com/assets/5567574/15344798/a4dafb2c-1c77-11e6-9ec5-4ebbd7cb7eb0.JPG)

Example file is here:
https://www.dropbox.com/s/qb93j78f8d22zj3/PET_Microclimate_Map2.gh?dl=0

I should have made this microclimate map recipe component a while ago.  @stgeorges had put his code for calculating PET into Ladybug_Ladybug so many months ago in order to facilitate this.

Let use know if you have any thoughts or feedback.
-Chris

**Comment by stgeorges** [comment]
@TheodoreGalanos , @chriswmackey ,

Is it possible to get those beta components related with importing cfd wind values to grasshopper, please? If not publicly, then by email? Thank you in advance.

@Vyrwn 
> Correct me if I am missing something but we could not run the analysis for the whole grid using the Ladybug_Thermal_Comfort_Indices component. We had to "trick" the component to do so, using the number of the grid points instead of 8760 hours. 

It's certainly is. You just had to define the particular ```HOY_``` for which you would like to conduct your analysis.

**Comment by Vyrwn** [comment]
@stgeorges 

First of all thank you very much for these components they are very much appreciated and useful. The issue is that creating a map as Chris is showing above, is not possible by defining the HOY_ if we want to test more than one grid point. We did not have the component above before so we did a work-around by adding a Grid_Point input that if !=none then it overrides the hours of the year to the number of the grid points and runs a "fake annual" simulation instead.

@chriswmackey thank you very much for the link, I will have a go to see how it works.

-Byron

**Comment by stgeorges** [comment]
Hi @Vyrwn ,
I would be very grateful if you could send me the particular .gh definition which replicates the situation you are describing. Thank you.

**Comment by Vyrwn** [comment]
Hi @stgeorges ,

I created a file with a small grid just to show what my issue was.

https://www.dropbox.com/s/ciys1spt6prqy0l/PET_on_Grid.gh?dl=0

I use generic data for Temp, Humidity and Wind just to make it simple. Let me know your thoughts and if there is something I am doing wrong.

-Byron

**Comment by mostaphaRoudsari** [comment]
@stgeorges you can find the files in [this discussion](http://www.grasshopper3d.com/group/ladybug/forum/topics/import-cfd-result-to-honeybee). They are both parsing the .csv file.

**Comment by chriswmackey** [comment]
@Vyrwn ,

Glad that you find the PET recipe component useful and I agree that you should be able to use the Thermal Indices component to make simple outdoor comfort maps if you are not interested in accounting for surface temperatures [similar to this simple UTCI map](http://hydrashare.github.io/hydra/viewer?owner=chriswmackey&fork=hydra_2&id=Urban_Microclimate_-_Simple_Spatial_UTCI&slide=1&scale=1&offset=0,0).

We will find some way to make this possible.

-Chris

**Comment by Vyrwn** [comment]
@chriswmackey,

Yes I agree that mapping PET or other indices on a grid is quite important and a much easier way of conveying the message. The process we follow is exactly the same as Theodore and you are describing above. Get the wind data from OpenFoam, then run radiation on the same grid points and then calculate comfort with some slight modifications that we did and we are still looking into.

As I said before, the main issue was the simulation of all grid points at the same time. For this I modified the component which you can find in this link. (the modification is not very scientific, but it does the trick).

byron-master/Ladybug_Thermal Comfort Indices_Grid.ghuser

I hope this is useful

-Byron

**Comment by stgeorges** [comment]
@TheodoreGalanos , @chriswmackey 
1) What does the ```WindFactor(10)``` column in the .csv file represent?
2) For what direction was the cfd simulation in openfoam been conducted? Westerly wind direction?

@mostaphaRoudsari 
Thank you for pointing me to the topic. I am a bit ashamed that I am not investing a little bit of my free time to check for the new topics on forum. I even missed this topic on github.

> I agree that you should be able to use the Thermal Indices component to make simple outdoor comfort maps if you are not interested in accounting for surface temperatures similar to this simple UTCI map.

One can always supply new values to the ```meanRadiantTemperature_``` input, which will account for the surface temperatures.

@Vyrwn Thank you for posting your .gh file.
Hi Byron,
I think I overcrowded this topic a bit. I opened a new [topic on forum](http://www.grasshopper3d.com/group/ladybug/forum/topics/pet-example-in-ladybug), with the reply to your PET issue. Thank you for being patience. In future, it is the best to seek help on forum, as it has better chances of getting noticed, than in here. I wouldn't even see your question in here, if Chris hasn't tagged me.

**Comment by stgeorges** [comment]
Hi @chriswmackey , @mostaphaRoudsari , @TheodoreGalanos ,

Is it possible to get an answer on these two questions related with the [.csv Open Foam result file from the grasshopper forum](http://www.grasshopper3d.com/xn/detail/2985220:Comment:1505794):

> 1) What does the WindFactor(10) column in the .csv file represent?
> 2) For what direction was the cfd simulation in openfoam been conducted? Westerly wind direction?

Thank you.
